pkg/distributors/contracts/RewardsScheduler.sol

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/SafeERC20.sol";
+import "./interfaces/IMultiRewards.sol";
+
+// solhint-disable not-rely-on-time
+
+/**
+ * Scheduler for MultiRewards contract
+ */
+contract RewardsScheduler {
+    using SafeERC20 for IERC20;
+
+    IMultiRewards private immutable _multirewards;
+
+    constructor() {
+        _multirewards = IMultiRewards(msg.sender);
+    }
+
+    enum RewardStatus { PENDING, STARTED }
+
+    struct ScheduledReward {
+        IERC20 pool;
+        IERC20 rewardsToken;
+        uint256 startTime;
+        address rewarder;
+        uint256 amount;
+        RewardStatus status;
+    }
+
+    event RewardScheduled(
+        bytes32 rewardId,
+        address indexed scheduler,
+        address indexed pool,
+        address indexed rewardsToken,
+        uint256 startTime,
+        uint256 amount
+    );
+    event RewardStarted(
+        bytes32 rewardId,
+        address indexed scheduler,
+        address indexed pool,
+        address indexed rewardsToken,
+        uint256 startTime,
+        uint256 amount
+    );
+
+    mapping(bytes32 => ScheduledReward) private _rewards;
+
+    function getScheduledRewardInfo(bytes32 rewardId) external view returns (ScheduledReward memory reward) {
+        return _rewards[rewardId];
+    }
+
+    function startRewards(bytes32[] calldata rewardIds) external {
+        for (uint256 r; r < rewardIds.length; r++) {
+            bytes32 rewardId = rewardIds[r];
+            ScheduledReward memory scheduledReward = _rewards[rewardId];
+
+            require(scheduledReward.startTime != 0, "reward has not been created");
+            require(scheduledReward.startTime <= block.timestamp, "reward cannot be started");
+            require(scheduledReward.status == RewardStatus.PENDING, "scheduled reward already activated");
+
+            scheduledReward.rewardsToken.approve(address(_multirewards), scheduledReward.amount);
```

Comment:
Agreed that there is no risk to granting infinite approval to the MultiRewards contract as it should only be able to deposit tokens to it's vault balance as part of starting a scheduled reward.

We _do_ hold token balances on the `RewardScheduler` - there is a safeTransferFrom the rewarder when you schedule a reward.

I'm considering it... find it a bit harder to read but yes, it would save gas for those poking this contract to start the reward, so it's probably worth it.   Would it make sense to do that approval when the reward is scheduled to shift the gas burden from the contract pokers to the reward schedulers?

Reply:
Ah yes, I got a bit mixed up on it holding balances. Besides I think the fact that it's approving a single known contract which we trust means it's an acceptable risk.

re moving approval from poker to scheduler: It's not a particularly gas intensive step and I'd prefer the readability improvements from keeping the approval and the call which requires it close to each other.

Reply:
Is this the only pending item on this PR?